src/nav_channel_ctrl.cpp

```diff
@@ -0,0 +1,259 @@
+
+        task_goal_position_.publish(goal_pos);
+    }
+
+    geometry_msgs::PoseStamped findMidpoint(int gate) {
+        // validates positions of props, then calculates midpoint, and returns
+        // it as PoseStamped
+        geometry_msgs::PoseStamped midpoint;
+
+        bool red = false;   // use two booleans to determine if props exist
+        bool green = false;
+        prop_mapper::Prop red_prop;
+        prop_mapper::Prop green_prop;
+        ROS_DEBUG_STREAM("Gate #" << gate);
+
+        for (int i = 0; (!red || !green) || i > sizeof(props.props) ; i++) {
+            double dist_to_gate = sqrt(pow(props.props[i].vector.x - current_pos.pose.pose.position.x, 2) + pow(props.props[i].vector.y - current_pos.pose.pose.position.y, 2));
+            if ((gate == 1 && dist_to_gate <= gate_max_dist) || (gate == 2 && dist_to_gate > gate_max_dist)) {
+            // using 8m as roughly 25ft
+            // gate 1 should be within 25ft, gate 2 should be at least 25ft away
+                if (props.props[i].prop_label == "Red Prop") {
+                    if (green) {
+                        float dist = sqrt(pow(green_prop.vector.x - props.props[i].vector.x, 2) + pow(green_prop.vector.y - props.props[i].vector.y, 2));
+                        if (dist < gate_max_width) {
+                            red = true;
+                            red_prop  = props.props[i];
+                        }
+                    }
+                    else {
+                        red = true;
+                        red_prop  = props.props[i];
+                    }
+                }
+                if (props.props[i].prop_label == "Green Prop") {
+                    if (red) {
+                        float dist = sqrt(pow(red_prop.vector.x - props.props[i].vector.x, 2) + pow(red_prop.vector.y - props.props[i].vector.y, 2));
+                        if (dist < gate_max_width) {
+                            green = true;
+                            green_prop = props.props[i];
+                        }
+                    }
+                    else {
+                        green = true;
+                        green_prop  = props.props[i];
+                    }
+                }
+            }
+            
+            
+        }
+
+        if (red && green) {
+            float red_x = red_prop.vector.x;
+            float red_y = red_prop.vector.y;
+            //float red_z = red_prop.vector.z;
+            float green_x = green_prop.vector.x;
+            float green_y = green_prop.vector.y;
+            //float green_z = green_prop.vector.z;
+
+            midpoint.pose.position.x = (red_x+green_x)/2;
+            midpoint.pose.position.y = (red_y+green_y)/2;
+            //midpoint.pose.position.z = (red_z+green_z)/2;
+            midpoint.pose.position.z = 0;
+        }
+
+        else {
+            ROS_DEBUG("gates were not found.");
```

Comment:
"Gates were not found" and returning midpoint could be INFO rather than debug. I typically use DEBUG for either messages that will show up on terminal repeatedly or messages that aren't really that important after a bug is resolved. But it's at your discretion!